File: trunk/managed/Warp3D/Warp3DVB/Warp_Vector.vb
' Created by: X
' http://www.createdbyx.com/
' ----------------------
' C# to VB.NET conversion of this code was done by hand and with the assistance of the 
' Convert C# to VB .NET application availible at http://www.kamalpatel.net/ConvertCSharp2VB.aspx
' ConvertCSharp2VB utility was developed by: Kamal Patel (http://www.KamalPatel.net) 
' ----------------------
' Thanks to Alan Simes ( http://alansimes.blogdns.net/ ) for creating the Warp3D library.
' http://alansimes.blogdns.net/cs/files/default.aspx
' ----------------------

Imports System

Namespace Rednettle.Warp3D
    Public Class warp_Vector
        Public x As Single = 0  'Cartesian (default)
        Public y As Single = 0  'Cartesian (default)
        Public z As Single = 0  'Cartesian (default),Cylindric
        Public r As Single = 0  'Cylindric
        Public theta As Single = 0  'Cylindric

        Public Sub New()
        End Sub

        Public Sub New(ByVal xpos As Single, ByVal ypos As Single, ByVal zpos As Single)
            x = xpos
            y = ypos
            z = zpos
        End Sub

        Public Function normalize() As warp_Vector
            Dim dist As Single = length()
            If dist = 0 Then Return Me
            Dim invdist As Single = 1 / dist
            x *= invdist
            y *= invdist
            z *= invdist
            Return Me
        End Function

        Public Function reverse() As warp_Vector
            x = -x
            y = -y
            z = -z
            Return Me
        End Function

        Public Function length() As Single
            Return CType(Math.Sqrt(x * x + y * y + z * z), Single)
        End Function

        Public Function transform(ByVal m As warp_Matrix) As warp_Vector
            Dim Newx As Single = x * m.m00 + y * m.m01 + z * m.m02 + m.m03
            Dim Newy As Single = x * m.m10 + y * m.m11 + z * m.m12 + m.m13
            Dim Newz As Single = x * m.m20 + y * m.m21 + z * m.m22 + m.m23

            Return New warp_Vector(Newx, Newy, Newz)
        End Function

        Public Sub buildCylindric()
            r = CType(Math.Sqrt(x * x + y * y), Single)
            theta = CType(Math.Atan2(x, y), Single)
        End Sub

        Public Sub buildCartesian()
            x = r * warp_Math.cos(theta)
            y = r * warp_Math.sin(theta)
        End Sub

        Public Shared Function getNormal(ByVal a As warp_Vector, ByVal b As warp_Vector) As warp_Vector
            Return vectorProduct(a, b).normalize()
        End Function

        Public Shared Function getNormal(ByVal a As warp_Vector, ByVal b As warp_Vector, ByVal c As warp_Vector) As warp_Vector
            Return vectorProduct(a, b, c).normalize()
        End Function

        Public Shared Function vectorProduct(ByVal a As warp_Vector, ByVal b As warp_Vector) As warp_Vector
            Return New warp_Vector(a.y * b.z - b.y * a.z, a.z * b.x - b.z * a.x, a.x * b.y - b.x * a.y)
        End Function

        Public Shared Function vectorProduct(ByVal a As warp_Vector, ByVal b As warp_Vector, ByVal c As warp_Vector) As warp_Vector
            Return vectorProduct(warp_Vector.sub(b, a), warp_Vector.sub(c, a))
        End Function

        Public Shared Function angle(ByVal a As warp_Vector, ByVal b As warp_Vector) As Single
            a.normalize()
            b.normalize()
            Return (a.x * b.x + a.y * b.y + a.z * b.z)
        End Function

        Public Shared Function add(ByVal a As warp_Vector, ByVal b As warp_Vector) As warp_Vector
            Return New warp_Vector(a.x + b.x, a.y + b.y, a.z + b.z)
        End Function

        Public Shared Function [sub](ByVal a As warp_Vector, ByVal b As warp_Vector) As warp_Vector
            Return New warp_Vector(a.x - b.x, a.y - b.y, a.z - b.z)
        End Function

        Public Shared Function scale(ByVal f As Single, ByVal a As warp_Vector) As warp_Vector
            Return New warp_Vector(f * a.x, f * a.y, f * a.z)
        End Function

        Public Shared Function len(ByVal a As warp_Vector) As Single
            Return CType(Math.Sqrt(a.x * a.x + a.y * a.y + a.z * a.z), Single)
        End Function
        ' 		
        ' 		Public static  void New random(single fact)
        ' 			' returns a random vector
        ' 		{
        ' 			Return New warp_Vector(fact*warp_Math.random(),fact*warp_Math.random(),fact*warp_Math.random())
        ' 		}
        '  
        ' 		Public String toString()
        ' 		{
        ' 			Return New String ("<vector x="+x+" y="+y+" z="+z+">\r\n")
        ' 		}
        ' */

        Public Function getClone() As warp_Vector
            Return New warp_Vector(x, y, z)
        End Function
    End Class
End Namespace

File: trunk/managed/Warp3Dmod/warp_Vector.cs
using System;

namespace Warp3D
{
	public class warp_Vector
	{
		public float x=0;      //Cartesian (default)
		public float y=0;      //Cartesian (default)
		public float z=0;      //Cartesian (default),Cylindric
		public float r=0;      //Cylindric
		public float theta=0;  //Cylindric

		public warp_Vector()
		{
		}
		
		public warp_Vector (float xpos, float ypos, float zpos)
		{
			x=xpos;
			y=ypos;
			z=zpos;
		}
		public warp_Vector normalize()
			// Normalizes the vector
		{
			float dist=length();
			if (dist==0) return this;
			float invdist=1/dist;
			x*=invdist;
			y*=invdist;
			z*=invdist;
			return this;
		}

		public warp_Vector reverse()
			// Reverses the vector
		{
			x=-x;
			y=-y;
			z=-z;
			return this;
		}

		public float length()
			// Lenght of this vector
		{
			return (float)Math.Sqrt(x*x+y*y+z*z);
		}

		public warp_Vector transform(warp_Matrix m)
			// Modifies the vector by matrix m
		{
			float newx = x*m.m00 + y*m.m01 + z*m.m02+ m.m03;
			float newy = x*m.m10 + y*m.m11 + z*m.m12+ m.m13;
			float newz = x*m.m20 + y*m.m21 + z*m.m22+ m.m23;

			return new warp_Vector(newx,newy,newz);
		}

		public void buildCylindric()
			// Builds the cylindric coordinates out of the given cartesian coordinates
		{
			r=(float)Math.Sqrt(x*x+y*y);
			theta=(float)Math.Atan2(x,y);
		}

		public void buildCartesian()
			// Builds the cartesian coordinates out of the given cylindric coordinates
		{
			x=r*warp_Math.cos(theta);
			y=r*warp_Math.sin(theta);
		}

		public static warp_Vector getNormal(warp_Vector a, warp_Vector b)
			// returns the normal vector of the plane defined by the two vectors
		{
			return vectorProduct(a,b).normalize();
		}

		public static warp_Vector getNormal(warp_Vector a, warp_Vector b, warp_Vector c)
			// returns the normal vector of the plane defined by the two vectors
		{
			return vectorProduct(a,b,c).normalize();
		}

		public static warp_Vector vectorProduct(warp_Vector a, warp_Vector b)
			// returns a x b
		{
			return new warp_Vector(a.y*b.z-b.y*a.z,a.z*b.x-b.z*a.x,a.x*b.y-b.x*a.y);
		}

		public static warp_Vector vectorProduct(warp_Vector a, warp_Vector b, warp_Vector c)
			// returns (b-a) x (c-a)
		{
			return vectorProduct(sub(b,a),sub(c,a));
		}

		public static float angle(warp_Vector a, warp_Vector b)
			// returns the angle between 2 vectors
		{
			a.normalize();
			b.normalize();
			return (a.x*b.x+a.y*b.y+a.z*b.z);
		}

		public static warp_Vector add(warp_Vector a, warp_Vector b)
			// adds 2 vectors
		{
			return new warp_Vector(a.x+b.x,a.y+b.y,a.z+b.z);
		}

		public static warp_Vector sub(warp_Vector a, warp_Vector b)
			// substracts 2 vectors
		{
			return new warp_Vector(a.x-b.x,a.y-b.y,a.z-b.z);
		}

		public static warp_Vector scale(float f, warp_Vector a)
			// substracts 2 vectors
		{
			return new warp_Vector(f*a.x,f*a.y,f*a.z);
		}

		public static float len(warp_Vector a)
			// length of vector
		{
			return (float)Math.Sqrt(a.x*a.x+a.y*a.y+a.z*a.z);
		}

		/*
		public static warp_Vector random(float fact)
			// returns a random vector
		{
			return new warp_Vector(fact*warp_Math.random(),fact*warp_Math.random(),fact*warp_Math.random());
		}

		public String toString()
		{
			return new String ("<vector x="+x+" y="+y+" z="+z+">\r\n");
		}
*/

		public warp_Vector getClone()
		{
			return new warp_Vector(x,y,z);
		}
	}
}
